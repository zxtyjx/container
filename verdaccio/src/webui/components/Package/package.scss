@import '../../styles/variable';

.package {
  .header {
    display: flex;
    align-items: center;
    margin: 10px 0 0;
  }

  .footer {
    display: flex;
    p.description {
      width: 100%;
      margin-bottom: 0;
      font-size: 14px;
      color: darkgray;
    }
  }

  .details {
    display: flex;
    font-size: 80%;
    color: $description_color;
    padding-top: 5px;
    .license {
      width: 20%;
      text-align: right;
    }

    .homepage {
      width: 80%;
    }
  }

  > a {
    display: block;
    position: relative;
    color: inherit;
    margin: 0;
    padding: 10px 0;
    cursor: pointer;
    text-decoration: none;

    .title {
      width: 100%;

      h1 {
        font-size: 18px;
        margin: 0;

        :global {
          .el-tag {
            margin-left: 5px;
          }
        }
      }
    }

    .author {
      color: $description_color;
      font-size: inherit;
      width: 30%;
      text-align: right;
    }

    &:hover {
      &::before {
        display: block;
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        background: rgba(255, 255, 255, 0.7);
        content: '';
      }

      &::after {
        display: block;
        position: absolute;
        top: 50%;
        transform: translateY(-50%);
        right: 0;
        left: 0;
        text-align: center;
        font-size: 18px;
      }
    }
  }
}
