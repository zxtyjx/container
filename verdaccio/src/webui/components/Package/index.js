import React from 'react';
import PropTypes from 'prop-types';
import {Tag} from 'element-react';
import {Link} from 'react-router-dom';

import {formatDateDistance} from '../../utils/package';

import classes from './package.scss';

const Package = ({name, version, author, description, license, time}) => {
  return (<section className={classes.package}>
    <Link to={`detail/${name}`}>
      <div className={classes.header}>
        <div className={classes.title}>
          <h1>
            {name} <Tag type="gray">v{version}</Tag>
          </h1>
        </div>
        <div role="author" className={classes.author}>
        { author ? `By: ${author}`: ''}
        </div>
      </div>
      <div className={classes.footer}>
        <p className={classes.description}>
          {description}
        </p>
      </div>
      <div className={classes.details}>
        <div className={classes.homepage}>
          {time ? `Published ${formatDateDistance(time)} ago` : ''}
        </div>
        <div className={classes.license}>
          {license}
        </div>
      </div>
    </Link>
  </section>);
};

Package.propTypes = {
  name: PropTypes.string,
  version: PropTypes.string,
  author: PropTypes.string,
  description: PropTypes.string,
  license: PropTypes.string,
  time: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
  ])
};

export default Package;
