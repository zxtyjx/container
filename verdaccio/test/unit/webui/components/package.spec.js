/**
 * Package component
 */

import React from 'react';
import { mount } from 'enzyme';
import Package from '../../../../src/webui/components/Package/index';
import { BrowserRouter } from 'react-router-dom';

/**
 * Generates one month back date from current time
 * @return {object} date object
 */
const dateOneMonthAgo = () => {
  const date = new Date();
  date.setMonth(date.getMonth() - 1);
  return date;
}

describe('<Package /> component', () => {
  it('should load the component', () => {
    const props = {
      name: 'verdaccio',
      version: '1.0.0',
      time: dateOneMonthAgo(),
      license: 'MIT',
      description: 'Private NPM repository',
      author: 'Sam'
    };
    const wrapper = mount(
      <BrowserRouter>
        <Package {...props}/>
      </BrowserRouter>
    );

    // integration expectations
    expect(wrapper.find('a').prop('href')).toEqual('detail/verdaccio');
    expect(wrapper.find('h1').text()).toEqual('verdaccio v1.0.0');
    expect(wrapper.find('.el-tag--gray').text()).toEqual('v1.0.0');
    expect(
      wrapper.find('div').filterWhere(n => n.prop('role') === 'author')
        .text()
    ).toEqual('By: Sam');
    expect(wrapper.find('p').text()).toEqual('Private NPM repository');
    expect(wrapper.find('.license').text()).toMatch(/MIT/);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should load the component without author', () => {
    const props = {
      name: 'verdaccio',
      version: '1.0.0',
      time: dateOneMonthAgo(),
      license: 'MIT',
      description: 'Private NPM repository'
    };
    const wrapper = mount(
      <BrowserRouter>
        <Package {...props} />
      </BrowserRouter>
    );

    // integration expectations
    expect(
      wrapper.find('div').filterWhere(n => n.prop('role') === 'author')
        .text()
    ).toEqual('');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
 