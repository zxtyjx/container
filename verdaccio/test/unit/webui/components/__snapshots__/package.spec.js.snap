// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Package /> component should load the component 1`] = `"<section class=\\"package\\"><a href=\\"detail/verdaccio\\"><div class=\\"header\\"><div class=\\"title\\"><h1>verdaccio <span class=\\"el-tag el-tag--gray\\">v1.0.0</span></h1></div><div role=\\"author\\" class=\\"author\\">By: Sam</div></div><div class=\\"footer\\"><p class=\\"description\\">Private NPM repository</p></div><div class=\\"details\\"><div class=\\"homepage\\">Published about 1 month ago</div><div class=\\"license\\">MIT</div></div></a></section>"`;

exports[`<Package /> component should load the component without author 1`] = `"<section class=\\"package\\"><a href=\\"detail/verdaccio\\"><div class=\\"header\\"><div class=\\"title\\"><h1>verdaccio <span class=\\"el-tag el-tag--gray\\">v1.0.0</span></h1></div><div role=\\"author\\" class=\\"author\\"></div></div><div class=\\"footer\\"><p class=\\"description\\">Private NPM repository</p></div><div class=\\"details\\"><div class=\\"homepage\\">Published about 1 month ago</div><div class=\\"license\\">MIT</div></div></a></section>"`;
